core/dataset.cc

```diff
@@ -1617,7 +1617,9 @@ generateRowsWhere(const SqlBindingScope & scope,
     UnboundEntities unbound = where.getUnbound();
 
     // Look for a free variable
-    bool needsColumns = unbound.hasUnboundVariables();
+    bool needsColumns = unbound.hasUnboundVariables() ||
+        unbound.funcs.find("columnCount") != unbound.funcs.end();
```

Comment:
This fix feels like a band-aid to me.  We should have an unbound.needsRow() function that looks for hasUnboundVariables() || hasRowFunction(), and implement those in some kind of a generic manner.